zerver/actions/message_send.py

```diff
@@ -257,13 +273,21 @@ def notification_recipients(setting: str) -> Set[int]:
         stream_push_user_ids = notification_recipients("push_notifications")
         stream_email_user_ids = notification_recipients("email_notifications")
 
-        if possible_stream_wildcard_mention:
-            # We calculate `stream_wildcard_mention_user_ids` only if there's a possible
-            # stream wildcard mention in the message. This is important so as to avoid
-            # unnecessarily sending huge user ID lists with thousands of elements
+        if possible_stream_wildcard_mention or possible_topic_wildcard_mention:
+            # We calculate `wildcard_mentions_notify_user_ids` only if there's a possible
+            # stream or topic wildcard mention in the message. This is important so as to
+            # avoid unnecessarily sending huge user ID lists with thousands of elements
             # to the event queue (which can happen because this setting is `True`
             # by default for new users.)
-            stream_wildcard_mention_user_ids = notification_recipients("wildcard_mentions_notify")
+            wildcard_mentions_notify_user_ids = notification_recipients("wildcard_mentions_notify")
+
+        if possible_stream_wildcard_mention:
+            stream_wildcard_mention_user_ids = wildcard_mentions_notify_user_ids
+
+        if possible_topic_wildcard_mention:
+            for user_id in wildcard_mentions_notify_user_ids:
+                if user_id in topic_participant_user_ids:
+                    topic_wildcard_mention_user_ids.add(user_id)
```

Comment:
I feel like this probably should be written using the Python set intersection syntax that probably takes care of doing the loop order starting from the smallest set for us, but if we're doing this loop structure you definitely want the other loop order, because `wildcard_mentions_notify_user_ids` is probably everyone subscribed to the stream, whereas `topic_wildcard_mention_user_ids` is typically going to be the small subset who've participated in this topic.

Reply:
>  this probably should be written using the Python set intersection syntax that probably takes care of doing the loop order starting from the smallest set for us

Yes, verified with [the docs](https://wiki.python.org/moin/TimeComplexity).

* Average Time Complexity for `intersection` is `O(min(len(setA), len(setB)))`
* Worst Time Complexity is `O(len(setA)*len(setB))` -- I think this would be the case only when accessing element from map is `O(n)`, which is rare.

In short, we have,
`topic_wildcard_mention_user_ids` = `topic_participant_user_ids.intersection(wildcard_mentions_notify_user_ids)`.